### ISSUE TITLE
Missing geometry in Ifc2x3tc1 file with IfcClosedShell

### ISSUE THREAD

**Issue by rubendel** [IfcOpenShell/IfcOpenShell#717]
There is an issue on the BIMserver issue tracker [1]. Some geometry is missing for the Ifc2x3 file [2].

No errors are reported, surprising fact is that the IFC4 file actually works correctly (but it uses a different geometry type).

Tested in v0.6.0 b4ce5be with layerset slicing and quantity calculation enabled.

[1] https://github.com/opensourceBIM/BIMserver/issues/1022
[2] [Balken_FGG.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/3825116/Balken_FGG.ifc.txt)

**Comment by Moult** [comment]
Tested the file in BlenderBIM and it works correctly?

![2019-11-10-133028_676x270_scrot](https://user-images.githubusercontent.com/88302/68537809-5c929000-03be-11ea-86ad-306f3ecd5180.png)

**Comment by rubendel** [comment]
Thanks, which version is that?

**Comment by aothms** [comment]
I can also not find an immediate issue with this file. There is a script to run the IfcGeomServer here https://raw.githubusercontent.com/IfcOpenShell/IfcOpenShell/v0.7.0/src/ifcopenshell-python/ifcopenshell/geom/client.py

and it outputs:

~~~
ENTITY:
id: 8443
guid: 3vsy5EawfAJh7Kbmk4KMz4
name: 1
type: IfcBuildingElementProxy
parentId: 123
matrix: [1.00000000e+00 0.00000000e+00 0.00000000e+00 3.56561895e+06
 0.00000000e+00 1.00000000e+00 0.00000000e+00 5.94108595e+06
 0.00000000e+00 0.00000000e+00 1.00000000e+00 0.00000000e+00
 0.00000000e+00 0.00000000e+00 0.00000000e+00 1.00000000e+00]
repId: 8418
positions: <bytes>
normals: <bytes>
indices: <bytes>
colors: <bytes>
materialIndices: <bytes>
extendedData: TOTAL_SURFACE_AREA: 3.14494033574554
              SURFACE_AREA_ALONG_X: 0.375975890930012
              SURFACE_AREA_ALONG_Y: 0.395843805364851
              SURFACE_AREA_ALONG_Z: 2.54245731407558
              BOUNDING_BOX_SIZE_ALONG_X: 5.89408950462341
              BOUNDING_BOX_SIZE_ALONG_Y: 2.95435920410161
              BOUNDING_BOX_SIZE_ALONG_Z: 0.05000000038147
              LARGEST_FACE_DIRECTION: [0, 0, -1]
              LARGEST_FACE_AREA: 0.142908237474653
~~~

**Comment by hlg** [comment]
Can you have another look at this? There must be an issue with the IfcGeomServer. I have run a series of conversions with modified versions of the offending IFC:

a) removed one face from the faceted BREP and
b) split the closed shell into two open shells.

These modifications were made in an attempt to reduce the number of faces as I was assuming the number of faces (1380) to be an issue, though it might be something totally different.

I found that your output above does also reflect the reduced geometry - not the full expected shell. In fact, the bounding box should be around 6.95 x 4.01 x 1.50 (around 8.0 x 1.5 x 1.5 but rotated), the largest face should be one of the two triangles at the top with direction [0,0,1]. There are 1380 triangular faces and I would expect the index buffer to be of size 4140, and the vertex buffer of size 12420. But the geom server doesn't produce these numbers for any of the files. Any ideas?

### Original file (same as yours above)

```
"TOTAL_SURFACE_AREA": 3.14494033574554,
"SURFACE_AREA_ALONG_X": 0.375975890930012,
"SURFACE_AREA_ALONG_Y": 0.395843805364851,
"SURFACE_AREA_ALONG_Z": 2.54245731407558,
"BOUNDING_BOX_SIZE_ALONG_X": 5.89408950462341,
"BOUNDING_BOX_SIZE_ALONG_Y": 2.95435920410161,
"BOUNDING_BOX_SIZE_ALONG_Z": 0.05000000038147,
"LARGEST_FACE_DIRECTION": [0,0,-1],
"LARGEST_FACE_AREA": 0.142908237474653

Vertex buffer size: 4230
Index buffer size: 1410
```

### One face removed from the faceted BREP (the closed shell not closed anymore)
[Balken_FGG_one-face-removed.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/6110573/Balken_FGG_one-face-removed.ifc.txt)
```
"TOTAL_SURFACE_AREA": 47.1722596548487,

"SURFACE_AREA_ALONG_X": 12.8463680630981,
"SURFACE_AREA_ALONG_Y": 21.6667205417289,
"SURFACE_AREA_ALONG_Z": 20.6994943167577,

"BOUNDING_BOX_SIZE_ALONG_X": 6.95193089447022,
"BOUNDING_BOX_SIZE_ALONG_Y": 4.01171791992191,
"BOUNDING_BOX_SIZE_ALONG_Z": 1.5000000114441,

"LARGEST_FACE_DIRECTION": [0,0,1],
"LARGEST_FACE_AREA": 5.17501443945967
}
Vertex buffer size:  10368
Index buffer size:  3456
```

### Split into two open shells (combined in an IfcShellBasedSurfaceModel)
[Balken_FGG_closed-shell-split.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/6110572/Balken_FGG_closed-shell-split.ifc.txt)
```
"TOTAL_SURFACE_AREA": 41.7571626027579,

"SURFACE_AREA_ALONG_X": 12.6824499276923,
"SURFACE_AREA_ALONG_Y": 21.5228448994927,
"SURFACE_AREA_ALONG_Z": 15.528694223164,

"BOUNDING_BOX_SIZE_ALONG_X": 6.95193089447022,
"BOUNDING_BOX_SIZE_ALONG_Y": 4.01171791992191,
"BOUNDING_BOX_SIZE_ALONG_Z": 1.5000000114441,

"LARGEST_FACE_DIRECTION": [0.38035684588176,-0.924839807637452,0],
"LARGEST_FACE_AREA": 5.17500166612441

Vertex buffer size:  9189
Index buffer size:  3063
```

**Comment by aothms** [comment]
I think I see where this is coming from. The setting in geom server client CALCULATE_QUANTITIES actually maps to the setting SEW_SHELLS in IfcOpenShell. If you disable CALCULATE_QUANTITIES quantities IfcOpenShell treats the faces as individual loose faces (more or less) which is fine for visualization. If you do use CALCULATE_QUANTITIES we 'sew' the faces into a manifold set, which solves for orientation issues so that it's reliable to calculate volume. 

This geometry is not valid though (geometrically or topologically) I don't see really what's going on:

![afbeelding](https://user-images.githubusercontent.com/1096535/110618708-4b66b880-8197-11eb-8de7-067d5acde485.png)

So when Open CASCADE's tries to orient it it apparently discards the non-manifold bits. There is logic in place that should collect the loose bits, but it isn't triggered in this case. 

I guess one thing we could do in create_solid_from_faces() is to count the number of faces during the execution of all the various steps. You can also just keep CALCULATE_QUANTITIES  off. It will save a lot of processing time for these detailed meshes.

**Comment by hlg** [comment]
Thanks, that's good insight and with CALCULATE_QUANTITIES off the display appears to be complete.

The attempt to sew faces could also explain the weird largest face orientation in the last sample where I have made an arbitrary split across the face set and likely left behind disconnected open shells. Now I get 1022 triangles, whereas before it had been 1152 with the one face removed, maybe the additional faces where also due to the sewing attempt?

I still don't have an idea what could be wrong with the original file, because I had manually converted the IFC to OBJ (using some regex/find/replace) and ran it through val3dity - with no errors. Maybe val3dity tolerances where set too high. I'm gonna try that again.

**Comment by hlg** [comment]
Here is the OBJ file: [Balken_FGG.obj.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/6115718/Balken_FGG.obj.txt).

Val3dity does indeed not report any errors with "snap tolerance" set to 0. I did also run FreeCAD's "Evaluate Mesh" tool: no flipped normals, duplicated faces or points, non-manifolds, degenerated faces, invalid indices, self-intersections or folds on surface. It reports the full number of faces (1380) and points (692), nothing discarded along the way.

**Comment by aothms** [comment]
IfcOpenShell does an initial vertex clustering to reduce the risk of errors in OCCT. You can also try reducing the Precision in the file as it will set the radius for clustering/

**Comment by hlg** [comment]
I see! That is the issue with the file then: the precision value is declared larger than the variance of coordinates. Works fine after adjusting the geometric representation context even with quantity calculation enabled.

Thanks for the help! You can close this issue.

**Comment by hlg** [comment]
Just as a follow-up. I guess that situations like the following will throw off the interpretation of the geometry, with clustering radius _p_ taken from an unfeasible IFC precision value.

![image](https://user-images.githubusercontent.com/356986/110772344-cabdc000-825b-11eb-867a-7d9e09929a88.png)

I don't immediately see an easy way to use these precision values when trying to cluster points and at the same time keep valid solids.

**Comment by aothms** [comment]
Indeed, vertex clustering is not topology preserving. The implementation I have is also not very robust wrt vertex order. It's just something in place so that OCCT doesn't crash on inconsistencies when these small details are fed to the boolean ops algorithms.

**Comment by BIMvoice** [comment]
Verified fixed on the current build (0.8.5-1c5b825): the original `Balken_FGG.ifc` now produces full beam geometry (692 vertices through the geometry iterator) — identical to the `one-face-removed` workaround variant, which confirms the `IfcClosedShell` handling no longer drops the shape. No errors, clean exit. Safe to close.
